airbyte-integrations/connectors/source-snowflake/src/main/java/io.airbyte.integrations.source.snowflake/SnowflakeDataSourceUtils.java

```diff
@@ -0,0 +1,197 @@
+          LOGGER.info("Authorization mode is 'Username and password'");
+          populateUsernamePasswordConfig(hikariConfig, config.get("credentials"));
+        }
+        default -> throw new IllegalArgumentException("Unrecognized auth type: " + authType);
+      }
+    } else {
+      LOGGER.info("Authorization mode is deprecated 'Username and password'. Please update your source configuration");
+      populateUsernamePasswordConfig(hikariConfig, config);
+    }
+
+    return new HikariDataSource(hikariConfig);
+  }
+
+  /**
+   * Method to make request for a new access token using refresh token and client credentials.
+   *
+   * @return access token
+   */
+  public static String getAccessTokenUsingRefreshToken(final String hostName,
+                                                       final String clientId,
+                                                       final String clientSecret,
+                                                       final String refreshCode)
+      throws IOException {
+    final var refreshTokenUri = String.format(REFRESH_TOKEN_URL, hostName);
+    final Map<String, String> requestBody = new HashMap<>();
+    requestBody.put("grant_type", "refresh_token");
+    requestBody.put("refresh_token", refreshCode);
+
+    try {
+      final BodyPublisher bodyPublisher = BodyPublishers.ofString(requestBody.keySet().stream()
+          .map(key -> key + "=" + URLEncoder.encode(requestBody.get(key), StandardCharsets.UTF_8))
+          .collect(joining("&")));
+
+      final byte[] authorization = Base64.getEncoder()
+          .encode((clientId + ":" + clientSecret).getBytes(StandardCharsets.UTF_8));
+
+      final HttpRequest request = HttpRequest.newBuilder()
+          .POST(bodyPublisher)
+          .uri(URI.create(refreshTokenUri))
+          .header("Content-Type", "application/x-www-form-urlencoded")
+          .header("Accept", "application/json")
+          .header("Authorization", "Basic " + new String(authorization, StandardCharsets.UTF_8))
+          .build();
+
+      final HttpResponse<String> response = httpClient.send(request,
+          HttpResponse.BodyHandlers.ofString());
+
+      final JsonNode jsonResponse = Jsons.deserialize(response.body());
+      if (jsonResponse.has("access_token")) {
+        return jsonResponse.get("access_token").asText();
+      } else {
+        throw new RuntimeException(
+            "Failed to obtain accessToken using refresh token. " + jsonResponse);
+      }
+    } catch (final InterruptedException e) {
+      throw new IOException("Failed to refreshToken", e);
+    }
+  }
+
+  public static String buildJDBCUrl(JsonNode config) {
+    final StringBuilder jdbcUrl = new StringBuilder(String.format("jdbc:snowflake://%s/?",
+        config.get("host").asText()));
+
+    // Add required properties
+    jdbcUrl.append(String.format(
+        "role=%s&warehouse=%s&database=%s&schema=%s&JDBC_QUERY_RESULT_FORMAT=%s&CLIENT_SESSION_KEEP_ALIVE=%s",
```

Comment:
there are several places where we're constructing a jdbc url manually; is it possible to use a `Map<String, String>` instead? (i.e. this parameter https://github.com/airbytehq/airbyte/blob/master/airbyte-db/lib/src/main/java/io/airbyte/db/Databases.java#L212)

Reply:
discussed over slack; going to keep using the URL params.